proto/spire/plugin/server/keymanager/v1/keymanager.proto

```diff
@@ -1,101 +1,130 @@
+    // NOT_FOUND is returned.
+    rpc GetPublicKey(GetPublicKeyRequest) returns (GetPublicKeyResponse);
 
-enum KeyType {
-    UNSPECIFIED_KEY_TYPE = 0;
-    EC_P256 = 1;
-    EC_P384 = 2;
-    RSA_1024 = 3;
-    RSA_2048 = 4;
-    RSA_4096 = 5;
-}
+    // Gets all public key information for the private keys managed by the
+    // plugin.
+    rpc GetPublicKeys(GetPublicKeysRequest) returns (GetPublicKeysResponse);
 
-enum HashAlgorithm {
-    UNSPECIFIED_HASH_ALGORITHM = 0;
-    // These entries (and their values) line up with a subset of the go
-    // crypto.Hash constants
-    SHA224 = 4;
-    SHA256 = 5;
-    SHA384 = 6;
-    SHA512 = 7;
-    SHA3_224 = 10;
-    SHA3_256 = 11;
-    SHA3_384 = 12;
-    SHA3_512 = 13;
-    SHA512_224 = 14;
-    SHA512_256 = 15;
+    // Signs data with the private key with the given ID and fingerprint. If a
+    // key with the given ID does not exist or has a different fingerprint,
+    // NOT_FOUND is returned.
+    rpc SignData(SignDataRequest) returns (SignDataResponse);
 }
 
-
 message PublicKey {
+    // Required. The ID of the key, as provided when the key was created.
     string id = 1;
+
+    // Required. The type of the key.
     KeyType type = 2;
+
+    // Required. The public key data (PKIX encoded).
     bytes pkix_data = 3;
+
+    // Required. Fingerprint of the public key. The (id,fingerprint) tuple
+    // represents a unique identifier for the key. The fingerprint is used to
+    // ensure that a signing operation is taking place with the intended key
+    // (see SignData). Plugins are welcome to choose their fingerprinting
+    // algorithm. A naive implementation is a hash over the PKIX data.
+    string fingerprint = 4;
 }
 
 message GenerateKeyRequest {
+    // Required. The ID to give the generated key (or to identify the existing
+    // key to overwrite (see GenerateKey).
     string key_id = 1;
+
+    // Required. The type of the key to generate.
     KeyType key_type = 2;
 }
 
 message GenerateKeyResponse {
+    // Required. The generated key.
     PublicKey public_key = 1;
 }
 
 message GetPublicKeyRequest {
+    // Required. The ID of the key to retrieve.
     string key_id = 1;
 }
 
 message GetPublicKeyResponse {
+    // Required. The public key to return.
     PublicKey public_key = 1;
 }
 
-
 message GetPublicKeysRequest {
 }
 
 message GetPublicKeysResponse {
+    // Required. The public keys managed by the KeyManager. May be empty.
     repeated PublicKey public_keys = 1;
 }
 
-message PSSOptions {
-    int32 salt_length = 1;
-    HashAlgorithm hash_algorithm = 2;
-}
-
 message SignDataRequest {
+    message PSSOptions {
+        // Required. The salt length.
+        int32 salt_length = 1;
+
+        // Required. The hash algorithm.
+        HashAlgorithm hash_algorithm = 2;
+    }
+
+    // Required. The ID of the key to use to sign the data.
     string key_id = 1;
+
+    // Required. The fingerprint of the key to use to sign the data.
+    string key_fingerprint = 2;
```

Comment:
Through offline conversation, and the reading of several comments, I've gathered that the keymanager gets to generate its own key fingerprints, and that they're relatively opaque to core, but there's no single place where that relationship is described. This will probably be something that keymanager plugins will be wondering about. Perhaps a small comment on the handling of this fingerprint (who generates it, when, what is it used for, what should I do with it), somewhere near the service definition or somewhere central, would be great